R/convertFlowCore.R

```diff
@@ -0,0 +1,177 @@
+
+      cov <- result$covar
+      colnames(cov) <- c(gate$xChannel, gate$yChannel)
+      rownames(cov) <- c(gate$xChannel, gate$yChannel)
+
+      return(flowCore::ellipsoidGate(filterId = gate$name, .gate = cov, mean = c(result$x, result$y)))
+    },
+    "PolygonGate" = {
+      m <- gate$model$polygon$vertices
+      if (is.null(dim(m))) {
+        m <- m[[1]]
+      }
+      colnames(m) <- c(gate$xChannel, gate$yChannel)
+      flowCore::polygonGate(filterId = gate$name, m)
+    },
+    "QuadrantGate" = stop("This gate has no representation in flowCore"),
+    "SplitGate" = stop("This gate has no representation in flowCore")
+  )
+}
+
+#' Scale a flowCore gate or flowDensity filter
+#'
+#' Applies a scale to points from a flowCore gate, so that it may
+#' be saved to CellEngine.
+#'
+#' @param gate The flowCore gate points (i.e. polygonGate@boundaries)
+#' @param scaleSet The CellEngine ScaleSet to apply.
+#' @export
+#' @examples
+#' \dontrun{
+#' sngl <- flowDensity(file, params)
+#' flowGate <- sngl@filter
+#' scaleSet <- getScaleSets("5d2f8b4b21fd0676fb3a6a8c")
+#' gatePoints <- scaleFlowCoreGate(flowGate, scaleSet)
+#' createPolygonGate(id, colnames(flowGate)[1], colnames(flowGate)[2], 'my gate', vertices=gatePoints)
+#' }
+scaleFlowCoreGate <- function (gate, scaleSet) {
+  xChannel <- colnames(gate)[1]
+  yChannel <- colnames(gate)[2]
+  scales <- data.frame(scaleSet$scales)
+  ss <- scales$scale
+  rownames(ss) <- scales$channelName
+
+  scaleX <- ss[xChannel,]
+  scaleY <- ss[yChannel,]
+
+  scaledGate = cbind(applyScale(scaleX, gate[,xChannel]), applyScale(scaleY, gate[,yChannel]))
+  colnames(scaledGate) <- c(xChannel, yChannel)
+  gate[,xChannel] <- unlist(scaledGate[,xChannel])
+  gate[,yChannel] <- unlist(scaledGate[,yChannel])
+
+  return(lapply(seq_len(nrow(gate)), function(i) gate[i,]))
+}
+
+scaleEllipseCenter <- function (center, scaleSet) {
+  xChannel <- names(center)[1]
+  yChannel <- names(center)[2]
+  scales <- data.frame(scaleSet$scales)
+  s <- scales$scale
+  rownames(s) <- scales$channelName
+
+  scaleX <- s[xChannel,]
+  scaleY <- s[yChannel,]
+
+  scaledCenter = cbind(applyScale(scaleX, center[xChannel]), applyScale(scaleY, center[yChannel]))
+  names(scaledCenter) <- c(xChannel, yChannel)
+  center[xChannel] <- unlist(scaledCenter[xChannel])
+  center[yChannel] <- unlist(scaledCenter[yChannel])
+
+  return(center)
+}
+
+#' Convert flowCore to gate
+#'
+#' Converts points in a flowCore gate or flowDensity filter to a CellEngine
+#' Gate and saves it to CellEngine. A ScaleSet must be provided to scale the
+#' gate correctly.
+#'
+#' @param flowObject A flowDensity object or a flowCore Gate
+#' @param scaleSet The CellEngine ScaleSet to be applied.
+#' @export
+#' @examples
+#' \dontrun{
+#' sngl <- flowDensity(file, params)
+#' scaleSet <- getScaleSets("5d2f8b4b21fd0676fb3a6a8c")
+#' convertFromFlowCore(sngl, scaleSet, name = "my gate")
+#' }
+convertFromFlowCore <- function (flowObject, scaleSet, name=NULL) {
```

Comment:
This needs to accept the other gate parameters (parentId, gid, maybe just use `...`).

Reply:
:+1: